pennylane/optimize/shot_adaptive.py

```diff
@@ -0,0 +1,449 @@
+        self.max_shots = None
+
+        # Running average of the parameter gradients
+        self.chi = None
+        # Running average of the variance of the parameter gradients
+        self.xi = None
+
+        super().__init__(stepsize=stepsize)
+
+    @staticmethod
+    def weighted_random_sampling(qnodes, coeffs, shots, argnums, *args, **kwargs):
+        """Returns an array containing length ``shots`` single-shot estimates
+        of the Hamiltonian gradient. The shots are distributed randomly over
+        the terms in the Hamiltonian, as per a Multinomial distribution.
+
+        Args:
+            qnodes (Sequence[.tape.QNode]): Sequence of QNodes, each one when evaluated
+                returning the corresponding expectation value of a term in the Hamiltonian.
+            coeffs (Sequence[float]): Sequences of coefficients corresponding to
+                each term in the Hamiltonian.
+            shots (int): The number of shots used to estimate the Hamiltonian expectation
+                value. These shots are distributed over the terms in the Hamiltonian,
+                as per a Multinomial distribution.
+            argnums (Sequence[int]): the QNode argument indices which are trainable
+            *args: Arguments to the QNodes
+            **kwargs: Keyword arguments to the QNodes
+
+        Returns:
+            array[float]: the single-shot gradients of the Hamiltonian expectation value
+        """
+
+        # determine the shot probability per term
+        prob_shots = np.abs(coeffs) / np.sum(np.abs(coeffs))
+
+        # construct the multinomial distribution, and sample
+        # from it to determine how many shots to apply per term
+        si = multinomial(n=shots, p=prob_shots)
+        shots_per_term = si.rvs()[0]
+
+        grads = []
+
+        for h, c, p, s in zip(qnodes, coeffs, prob_shots, shots_per_term):
+
+            # if the number of shots is 0, do nothing
+            if s == 0:
+                continue
+
+            # set the QNode device shots
+            h.device.shots = [(1, s)]
+
+            jacs = []
+            for i in argnums:
+                j = qml.jacobian(h, argnum=i)(*args, **kwargs)
+
+                if s == 1:
+                    j = np.expand_dims(j, 0)
+
+                # Divide each term by the probability per shot. This is
+                # because we are sampling one at a time.
+                jacs.append(c * j / p)
+
+            grads.append(jacs)
+
+        return [np.concatenate(i) for i in zip(*grads)]
+
+    @staticmethod
+    def check_device(dev):
+        r"""Verifies that the device used by the objective function is non-analytic.
+
+        Args:
+            dev (.Device): the device to verify
+
+        Raises:
+            ValueError: if the device is analytic
+        """
+        if dev.analytic:
+            raise ValueError("The Rosalin optimizer can only be used with non-analytic devices")

```

Comment:
We could give a hint that this means `shots` has to be an integer?

Reply:
How about the following?
```suggestion
            raise ValueError(
                "The Rosalin optimizer can only be used with non-analytic devices "
                "that estimate expectation values with a finite number of shots."
            )
```